ports/rp2/mpthreadport.c

```diff
@@ -45,16 +45,15 @@ STATIC uint32_t *core1_stack = NULL;
 STATIC size_t core1_stack_num_words = 0;
 
 // Thread mutex.
-STATIC mp_thread_mutex_t atomic_mutex;
+STATIC recursive_mutex_t atomic_mutex;
+STATIC unsigned int atomic_taken;
 
 uint32_t mp_thread_begin_atomic_section(void) {
     if (core1_entry) {
         // When both cores are executing, we also need to provide
         // full mutual exclusion.
-        mp_thread_mutex_lock(&atomic_mutex, 1);
-        // In case this atomic section is for flash access, then
-        // suspend the other core.
-        multicore_lockout_start_blocking();
+        recursive_mutex_enter_blocking(&atomic_mutex);
+        ++atomic_taken;
```

Comment:
Doesn't `recursive_mutex_enter_blocking()` provide this recursion count function?